pkg/agent/controller/traceflow/traceflow_controller.go

```diff
@@ -0,0 +1,404 @@
+	startTime := time.Now()
+	defer func() {
+		klog.V(4).Infof("Finished syncing Traceflow for %s. (%v)", traceflowName, time.Since(startTime))
+	}()
+
+	tf, err := c.traceflowLister.Get(traceflowName)
+	if err != nil {
+		return err
+	}
+	switch tf.Status.Phase {
+	case opsv1alpha1.Running:
+		if tf.Status.DataplaneTag != 0 {
+			start := false
+			c.runningTraceflowsMutex.Lock()
+			if _, ok := c.runningTraceflows[tf.Status.DataplaneTag]; !ok {
+				c.runningTraceflows[tf.Status.DataplaneTag] = tf.Name
+				start = true
+			}
+			c.runningTraceflowsMutex.Unlock()
+			if start {
+				err = c.startTraceflow(tf)
+			}
+		} else {
+			klog.Warningf("Invalid data plane tag %d for Traceflow %s", tf.Status.DataplaneTag, tf.Name)
+		}
+	default:
+		c.deallocateTag(tf)
+	}
+	return err
+}
+
+// startTraceflow deploys OVS flow entries for Traceflow and inject packet if current Node
+// is Sender Node.
+func (c *Controller) startTraceflow(tf *opsv1alpha1.Traceflow) error {
+	// Deploy flow entries for traceflow
+	klog.V(2).Infof("Deploy flow entries for Traceflow %s", tf.Name)
+	err := c.ofClient.InstallTraceflowFlows(tf.Status.DataplaneTag)
+	defer func() {
+		if err != nil {
+			c.errorTraceflowCRD(tf, fmt.Sprintf("Node: %s, error: %+v", tf.Name, err))
+		}
+	}()
+	if err != nil {
+		return err
+	}
+
+	// TODO: let controller compute the source Node, and the source Node can just return an error,
+	//  if fails to find the Pod.
+	// Inject packet if this Node is sender.
+	podInterfaces := c.interfaceStore.GetContainerInterfacesByPod(tf.Spec.Source.Pod, tf.Spec.Source.Namespace)
+	// Skip inject packet if Pod not found in current Node.
+	if len(podInterfaces) == 0 {
+		return nil
+	}
+	err = c.injectPacket(tf)
+	return err
+}
+
+func (c *Controller) injectPacket(tf *opsv1alpha1.Traceflow) error {
+	podInterfaces := c.interfaceStore.GetContainerInterfacesByPod(tf.Spec.Source.Pod, tf.Spec.Source.Namespace)
+	// Update Traceflow phase to Running.
+	klog.V(2).Infof("Injecting packet for Traceflow %s", tf.Name)
+	c.injectedTagsMutex.Lock()
+	c.injectedTags[tf.Status.DataplaneTag] = tf.Name
+	c.injectedTagsMutex.Unlock()
+
+	// Calculate destination MAC/IP.
+	dstMAC := ""
+	dstIP := tf.Spec.Destination.IP
+	isInterNode := true
+	// TODO: Find MAC by dstIP
+	if dstIP == "" {
+		dstPodInterfaces := c.interfaceStore.GetContainerInterfacesByPod(tf.Spec.Destination.Pod, tf.Spec.Destination.Namespace)
+		if len(dstPodInterfaces) > 0 {
+			dstMAC = dstPodInterfaces[0].MAC.String()
+			dstIP = dstPodInterfaces[0].IP.String()
+			isInterNode = false
+		} else {
+			dstPod, err := c.kubeClient.CoreV1().Pods(tf.Spec.Destination.Namespace).Get(context.TODO(), tf.Spec.Destination.Pod, v1.GetOptions{})
+			if err != nil {
+				return err
+			}
+			// dstMAC is "" here, will be set to Gateway MAC in ofClient.SendTraceflowPacket
+			dstIP = dstPod.Status.PodIP
+		}
+	}
+	if isInterNode {
+		if c.networkConfig.TunnelType != ovsconfig.GeneveTunnel {
```

Comment:
Since we have hybrid mode. Here you should call NeedsEncapToPeer() to see if tunnel is needed, if not we should return error too.